File: AgricultureApp/Services/FermeService.cs
// Services/FermeService.cs
using Microsoft.EntityFrameworkCore;
using AgricultureApp.Data;
using AgricultureApp.Models.DTOs;
using AgricultureApp.Models.Entities;

namespace AgricultureApp.Services
{
    public interface IFermeService
    {
        Task<List<FermeDto>> GetAllFermes();
        Task<List<FermeDto>> GetFermesByAgriculteur(int agriculteurId);
        Task<FermeDetailDto> GetFermeWithParcelles(int id);
        Task<FermeDto> CreateFerme(CreateFermeDto dto);
        Task<FermeDto> UpdateFerme(int id, UpdateFermeDto dto);
        Task<bool> DeleteFerme(int id);
        Task<bool> AssignerParcelles(int fermeId, List<int> parcelleIds);
        Task<bool> RetirerParcelle(int fermeId, int parcelleId);
        Task<List<ParcelleSimplifieeDto>> GetParcellesByFerme(int fermeId);
        Task<Parcelle> CreateParcelleDansFerme(int fermeId, DessinParcelleDto dto);
    }

    public class FermeService : IFermeService
    {
        private readonly ApplicationDbContext _context;

        public FermeService(ApplicationDbContext context)
        {
            _context = context;
        }

        public async Task<List<FermeDto>> GetAllFermes()
        {
            return await _context.Fermes
                .Include(f => f.Parcelles)
                .Select(f => new FermeDto
                {
                    Id = f.Id,
                    Nom = f.Nom,
                    AgriculteurId = f.AgriculteurId,
                    Gouvernorat = f.Gouvernorat,
                    Delegation = f.Delegation,
                    Secteur = f.Secteur,
                    Description = f.Description,
                    Couleur = f.Couleur,
                    NombreParcelles = f.Parcelles.Count,
                    SuperficieTotale = f.Parcelles.Sum(p => p.Surface),
                    DateCreation = f.DateCreation
                })
                .ToListAsync();
        }

        public async Task<List<FermeDto>> GetFermesByAgriculteur(int agriculteurId)
        {
            return await _context.Fermes
                .Include(f => f.Parcelles)
                .Where(f => f.AgriculteurId == agriculteurId)
                .Select(f => new FermeDto
                {
                    Id = f.Id,
                    Nom = f.Nom,
                    AgriculteurId = f.AgriculteurId,
                    Gouvernorat = f.Gouvernorat,
                    Delegation = f.Delegation,
                    Secteur = f.Secteur,
                    Description = f.Description,
                    Couleur = f.Couleur,
                    NombreParcelles = f.Parcelles.Count,
                    SuperficieTotale = f.Parcelles.Sum(p => p.Surface),
                    DateCreation = f.DateCreation
                })
                .ToListAsync();
        }

        public async Task<FermeDetailDto> GetFermeWithParcelles(int id)
        {
            var ferme = await _context.Fermes
                .Include(f => f.Parcelles)
                .FirstOrDefaultAsync(f => f.Id == id);

            if (ferme == null)
                return null;

            return new FermeDetailDto
            {
                Id = ferme.Id,
                Nom = ferme.Nom,
                AgriculteurId = ferme.AgriculteurId,
                Gouvernorat = ferme.Gouvernorat,
                Delegation = ferme.Delegation,
                Secteur = ferme.Secteur,
                Description = ferme.Description,
                Couleur = ferme.Couleur,
                NombreParcelles = ferme.Parcelles.Count,
                SuperficieTotale = ferme.Parcelles.Sum(p => p.Surface),
                DateCreation = ferme.DateCreation,
                Parcelles = ferme.Parcelles.Select(p => new ParcelleSimplifieeDto
                {
                    Id = p.Id,
                    Nom = p.Nom,
                    Surface = p.Surface,
                    Culture = p.Culture,
                    Couleur = p.Couleur,
                    EstSynchronise = p.EstSynchronise
                }).ToList()
            };
        }

        // Services/FermeService.cs
        public async Task<FermeDto> CreateFerme(CreateFermeDto dto)
        {
            try
            {
                Console.WriteLine("=== BACKEND - Création ferme ===");
                Console.WriteLine($"Nom: '{dto.Nom}'");
                Console.WriteLine($"AgriculteurId: {dto.AgriculteurId} (type: {dto.AgriculteurId.GetType()})");
                Console.WriteLine($"Gouvernorat: '{dto.Gouvernorat}'");
                Console.WriteLine($"Delegation: '{dto.Delegation}'");

                // Vérification explicite
                if (dto.AgriculteurId <= 0)
                {
                    throw new Exception("ID agriculteur invalide");
                }

                var ferme = new Ferme
                {
                    Nom = dto.Nom ?? "Ferme sans nom",
                    AgriculteurId = dto.AgriculteurId,
                    Gouvernorat = dto.Gouvernorat,
                    Delegation = dto.Delegation,
                    Secteur = dto.Secteur,
                    Description = dto.Description,
                    Couleur = dto.Couleur ?? "#4CAF50",
                    DateCreation = DateTime.UtcNow
                };

                _context.Fermes.Add(ferme);
                await _context.SaveChangesAsync();

                Console.WriteLine($"Ferme créée avec ID: {ferme.Id}");

                return new FermeDto
                {
                    Id = ferme.Id,
                    Nom = ferme.Nom,
                    AgriculteurId = ferme.AgriculteurId,
                    Gouvernorat = ferme.Gouvernorat,
                    Delegation = ferme.Delegation,
                    Secteur = ferme.Secteur,
                    Description = ferme.Description,
                    Couleur = ferme.Couleur,
                    NombreParcelles = 0,
                    SuperficieTotale = 0,
                    DateCreation = ferme.DateCreation
                };
            }
            catch (Exception ex)
            {
                Console.WriteLine($"ERREUR BACKEND: {ex.Message}");
                Console.WriteLine($"Stack trace: {ex.StackTrace}");
                throw;
            }
        }

        public async Task<FermeDto> UpdateFerme(int id, UpdateFermeDto dto)
        {
            var ferme = await _context.Fermes.FindAsync(id);
            if (ferme == null)
                return null;

            if (!string.IsNullOrEmpty(dto.Nom))
                ferme.Nom = dto.Nom;

            if (!string.IsNullOrEmpty(dto.Gouvernorat))
                ferme.Gouvernorat = dto.Gouvernorat;

            if (!string.IsNullOrEmpty(dto.Delegation))
                ferme.Delegation = dto.Delegation;

            if (!string.IsNullOrEmpty(dto.Secteur))
                ferme.Secteur = dto.Secteur;

            if (!string.IsNullOrEmpty(dto.Description))
                ferme.Description = dto.Description;

            if (!string.IsNullOrEmpty(dto.Couleur))
                ferme.Couleur = dto.Couleur;

            ferme.DateModification = DateTime.UtcNow;

            await _context.SaveChangesAsync();

            return new FermeDto
            {
                Id = ferme.Id,
                Nom = ferme.Nom,
                AgriculteurId = ferme.AgriculteurId,
                Gouvernorat = ferme.Gouvernorat,
                Delegation = ferme.Delegation,
                Secteur = ferme.Secteur,
                Description = ferme.Description,
                Couleur = ferme.Couleur,
                DateCreation = ferme.DateCreation
            };
        }

        public async Task<bool> DeleteFerme(int id)
        {
            var ferme = await _context.Fermes
                .Include(f => f.Parcelles)
                .FirstOrDefaultAsync(f => f.Id == id);

            if (ferme == null)
                return false;

            // Retirer l'association des parcelles avant de supprimer la ferme
            foreach (var parcelle in ferme.Parcelles)
            {
                parcelle.FermeId = null;
            }

            _context.Fermes.Remove(ferme);
            await _context.SaveChangesAsync();
            return true;
        }

        public async Task<bool> AssignerParcelles(int fermeId, List<int> parcelleIds)
        {
            var ferme = await _context.Fermes.FindAsync(fermeId);
            if (ferme == null)
                return false;

            var parcelles = await _context.Parcelles
                .Where(p => parcelleIds.Contains(p.Id) && p.AgriculteurId == ferme.AgriculteurId)
                .ToListAsync();

            foreach (var parcelle in parcelles)
            {
                parcelle.FermeId = fermeId;
            }

            await _context.SaveChangesAsync();
            return true;
        }

        public async Task<bool> RetirerParcelle(int fermeId, int parcelleId)
        {
            var parcelle = await _context.Parcelles
                .FirstOrDefaultAsync(p => p.Id == parcelleId && p.FermeId == fermeId);

            if (parcelle == null)
                return false;

            parcelle.FermeId = null;
            await _context.SaveChangesAsync();
            return true;
        }
        // Ajoutez ces méthodes dans FermeService.cs

        public async Task<Parcelle> CreateParcelleDansFerme(int fermeId, DessinParcelleDto dto)
        {
            var ferme = await _context.Fermes
                .Include(f => f.Parcelles)
                .FirstOrDefaultAsync(f => f.Id == fermeId);

            if (ferme == null)
                return null;

            var parcelle = new Parcelle
            {
                Nom = dto.Nom,
                Surface = dto.Surface,
                Culture = dto.Culture,
                Couleur = dto.Couleur,
                AgriculteurId = ferme.AgriculteurId,
                FermeId = fermeId,
                Latitude = dto.Latitude,
                Longitude = dto.Longitude,
                Geometrie = dto.Geometrie,
                Gouvernorat = dto.Gouvernorat ?? ferme.Gouvernorat,
                Delegation = dto.Delegation ?? ferme.Delegation,
                DateCreation = DateTime.UtcNow,
                EstSynchronise = true
            };

            _context.Parcelles.Add(parcelle);
            await _context.SaveChangesAsync();

            return parcelle;
        }

        public async Task<List<ParcelleSimplifieeDto>> GetParcellesByFerme(int fermeId)
        {
            return await _context.Parcelles
                .Where(p => p.FermeId == fermeId)
                .Select(p => new ParcelleSimplifieeDto
                {
                    Id = p.Id,
                    Nom = p.Nom,
                    Surface = p.Surface,
                    Culture = p.Culture,
                    Couleur = p.Couleur,
                    EstSynchronise = p.EstSynchronise
                })
                .ToListAsync();
        }
    }
}

File: Agriculture_App/src/app/services/api/ferme.service.ts
// services/api/ferme.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { map } from 'rxjs/operators';
import { Parcelle } from './parcelle.service';

export interface Ferme {
  id: number;
  nom: string;
  agriculteurId: number;
  localisation?: string;
  gouvernorat?: string;
  delegation?: string;
  superficieTotale: number;
  nombreParcelles: number;
  dateCreation: Date;
  description?: string;
  couleur?: string;
}

export interface FermeDetail extends Ferme {
  parcelles: ParcelleSimplifiee[];
}

export interface ParcelleSimplifiee {
  geometrie: any;
  id: number;
  nom: string;
  surface: number;
  culture?: string;
  couleur: string;
  estSynchronise: boolean;
  altitudeMin?: number;
  altitudeMax?: number;
  altitudeMoyenne?: number;
  penteMoyenne?: number;
  classePente?: string;
  exposition?: string;
}

@Injectable({
  providedIn: 'root'
})
export class FermeService {
  private apiUrl = 'http://localhost:5160/api/Fermes'; // À adapter selon votre API

  constructor(private http: HttpClient) {}

  // Récupérer toutes les fermes
  getAllFermes(): Observable<Ferme[]> {
    // Simulation - À remplacer par un vrai appel API
    return this.http.get<Ferme[]>(this.apiUrl).pipe(
      map(fermes => fermes.map(f => ({
        ...f,
        dateCreation: new Date(f.dateCreation)
      })))
    );
  }

  // Récupérer les fermes d'un agriculteur
  getFermesByAgriculteur(agriculteurId: number): Observable<Ferme[]> {
    return this.http.get<Ferme[]>(`${this.apiUrl}?agriculteurId=${agriculteurId}`).pipe(
      map(fermes => fermes.map(f => ({
        ...f,
        dateCreation: new Date(f.dateCreation)
      })))
    );
  }

  // Récupérer une ferme avec ses parcelles
  getFermeWithParcelles(fermeId: number): Observable<FermeDetail> {
    return this.http.get<FermeDetail>(`${this.apiUrl}/${fermeId}/details`).pipe(
      map(ferme => ({
        ...ferme,
        dateCreation: new Date(ferme.dateCreation)
      }))
    );
  }

  // Créer une nouvelle ferme
  createFerme(ferme: Partial<Ferme>): Observable<Ferme> {
    return this.http.post<Ferme>(this.apiUrl, ferme);
  }

  // Mettre à jour une ferme
  updateFerme(id: number, ferme: Partial<Ferme>): Observable<Ferme> {
    return this.http.put<Ferme>(`${this.apiUrl}/${id}`, ferme);
  }

  // Supprimer une ferme
  deleteFerme(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // Assigner des parcelles à une ferme
  assignerParcelles(fermeId: number, parcelleIds: number[]): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${fermeId}/parcelles/assigner`, { parcelleIds });
  }

  // Retirer une parcelle d'une ferme
  retirerParcelle(fermeId: number, parcelleId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${fermeId}/parcelles/${parcelleId}`);
  }

  getParcellesByFerme(fermeId: number): Observable<Parcelle[]> {
    return this.http.get<Parcelle[]>(`${this.apiUrl}/${fermeId}/parcelles`);
  }

// Ajouter une parcelle à une ferme
  ajouterParcelleAFerme(fermeId: number, parcelleId: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/${fermeId}/parcelles/${parcelleId}`, {});
  }

  // Créer une parcelle directement dans une ferme
  createParcelleDansFerme(fermeId: number, parcelleData: any): Observable<Parcelle> {
    return this.http.post<Parcelle>(`${this.apiUrl}/${fermeId}/parcelles/creer`, parcelleData);
  }
}
